### PR TITLE
Add options for reducing the framerate and muting the window when coopdx loses focus

### PR DESCRIPTION
**Pull request by tulpenkiste**
This PR adds:
- A new function to the window manager API, `bool (*has_focus)(void)` that checks if the window actively has focus (specifically checks for Keyboard Input on SDL)
- 2 new options in game
  - Reduce framerate on window focus lost (power saving, default: true)
  - Mute game audio on window focus lost (distraction-preventing, default: true)
  
The English text that I added for the new settings could probably be improved.
  
Let me know if theres a compilation/code issue for the DXGI and SDL1 window api backends.

Tested using SDL2 and DXGI.

### CODE REVIEW

File: lang/Czech.ini [hunk 1]
```diff
@@ -124,6 +124,7 @@ FORCE_4BY3 = "Vždy 4:3"
 PRELOAD_TEXTURES = "Načíst textury předem"
 VSYNC = "VSync"
 UNCAPPED_FRAMERATE = "Neomezené FPS"
+REDUCE_FRAMERATE_FOCUS_LOSS = "REDUCE_FRAMERATE_FOCUS_LOSS"
```
Comment (AgentXLP) [inline lang/Czech.ini]: Not translated

File: lang/Czech.ini [hunk 2]
```diff
@@ -381,6 +382,7 @@ MUSIC_VOLUME = "Hlasitost písničky"
 SFX_VOLUME = "Hlasitost zvuků"
 ENV_VOLUME = "Hlasitost prostředí"
 FADEOUT = "Ztišit daleké zvuky"
+MUTE_FOCUS_LOSS = "MUTE_FOCUS_LOSS"
```
Comment (AgentXLP) [inline lang/Czech.ini]: Not translated

File: lang/Dutch.ini [hunk 3]
```diff
@@ -124,6 +124,7 @@ FORCE_4BY3 = "Forceer 4:3"
 PRELOAD_TEXTURES = "Textures Voorladen"
 VSYNC = "VSync"
 UNCAPPED_FRAMERATE = "Ongelimiteerde Framerate"
+REDUCE_FRAMERATE_FOCUS_LOSS = "REDUCE_FRAMERATE_FOCUS_LOSS"
```
Comment (AgentXLP) [inline lang/Dutch.ini]: Not translated

File: lang/Dutch.ini [hunk 4]
```diff
@@ -381,6 +382,7 @@ MUSIC_VOLUME = "Muziek Geluid"
 SFX_VOLUME = "Sfx Geluid"
 ENV_VOLUME = "Env Geluid"
 FADEOUT = "Vervaagd Geluid in de Verte"
+MUTE_FOCUS_LOSS = "MUTE_FOCUS_LOSS"
```
Comment (AgentXLP) [inline lang/Dutch.ini]: Not translated

File: lang/Spanish.ini [hunk 5]
```diff
@@ -124,6 +124,7 @@ FORCE_4BY3 = "Forzar 4:3"
 PRELOAD_TEXTURES = "Precargar texturas"
 VSYNC = "Sincronización vertical"
 UNCAPPED_FRAMERATE = "FPS sin límite"
+REDUCE_FRAMERATE_FOCUS_LOSS = "Reducir la velocidad de fotogramas cuando la ventana no está enfocada"
```
Comment (AgentXLP) [inline lang/Spanish.ini]: Did you test to see if this fits in the panel?

File: lang/Spanish.ini [hunk 6]
```diff
@@ -381,6 +382,7 @@ MUSIC_VOLUME = "Volumen de Música"
 SFX_VOLUME = "Volumen de Efectos de Sonido"
 ENV_VOLUME = "Volumen de Entorno"
 FADEOUT = "Disminuir sonidos lejanos"
+MUTE_FOCUS_LOSS = "Silenciar el sonido cuando la ventana está desenfocada"
```
Comment (AgentXLP) [inline lang/Spanish.ini]: Did you test to see if this fits in the panel?

File: src/pc/gfx/gfx_dxgi.cpp [hunk 7]
```diff
@@ -685,6 +685,11 @@ HWND gfx_dxgi_get_h_wnd(void) {
 void gfx_dxgi_shutdown(void) {
 }
 
+// TODO: is this even right????
```
Comment (AgentXLP) [inline src/pc/gfx/gfx_dxgi.cpp]: Please test this

File: src/pc/gfx/gfx_sdl1.c [hunk 8]
```diff
@@ -197,6 +197,10 @@ static void gfx_sdl_shutdown(void) {
         SDL_Quit();
 }
 
+static bool gfx_sdl_has_focus(void) {
+    return SDL_GetAppState() & SDL_APPINPUTFOCUS;
```
Comment (AgentXLP) [inline src/pc/gfx/gfx_sdl1.c:201]: You tested this right?

File: src/pc/pc_main.c [hunk 9]
```diff
@@ -177,8 +177,14 @@ void produce_interpolation_frames_and_delay(void) {
         gfx_end_frame();
 
         // delay
-        if (!configUncappedFramerate) {
-            f64 targetDelta = 1.0 / (f64) configFrameLimit;
+        f64 tarFramerate = 0;
+
+        // TODO: ensure this is properly implemented on all window backends
```
Comment (AgentXLP) [inline src/pc/pc_main.c]: Please test

File: src/pc/pc_main.c [hunk 10]
```diff
@@ -177,8 +177,14 @@ void produce_interpolation_frames_and_delay(void) {
         gfx_end_frame();
 
         // delay
-        if (!configUncappedFramerate) {
-            f64 targetDelta = 1.0 / (f64) configFrameLimit;
+        f64 tarFramerate = 0;
```
Comment (AgentXLP) [inline src/pc/pc_main.c]: `tar`?

**Comment by Agent-11** [comment]
Also please retarget this PR to `dev.`

**Comment by tulpenkiste** [comment]
> Also please retarget this PR to `dev`.
Changed the target to the `dev` branch.

File: src/pc/gfx/gfx_sdl1.c (hunk 8 above)
Comment (retcinder) [reply src/pc/gfx/gfx_sdl1.c:201]: I haven't (due to compilation issues with SDL1 (caused by SDL_Scancode, which I have not touched)). However, according to the [libsdl 1.2 documentation](https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdlgetappstate.html), it is correct.

File: src/pc/gfx/gfx_dxgi.cpp (hunk 7 above)
Comment (retcinder) [reply src/pc/gfx/gfx_dxgi.cpp]: :+1: Will test it now.

File: src/pc/gfx/gfx_sdl1.c (hunk 8 above)
Comment (tulpenkiste) [reply src/pc/gfx/gfx_sdl1.c:201]: > I haven't (due to compilation issues with SDL1 (caused by SDL_Scancode, which I have not touched)).

Ignore this, I am just stupid and forgot how CONTROLLER_API works.

File: src/pc/gfx/gfx_dxgi.cpp (hunk 7 above)
Comment (retcinder) [reply src/pc/gfx/gfx_dxgi.cpp]: Tested on Windows 10, targeting D3D12. Works as intended, will be removing the comment.

File: src/pc/pc_main.c (hunk 10 above)
Comment (retcinder) [reply src/pc/pc_main.c]: `tar` meant target, albeit it was pretty unclear so I renamed the variable to `targetFramerate`.

File: src/pc/gfx/gfx_sdl1.c (hunk 8 above)
Comment (AgentXLP) [reply src/pc/gfx/gfx_sdl1.c:201]: Okay then

**Comment by tulpenkiste** [comment]
Converting to draft until I can make the translations fit

File: lang/Polish.ini [hunk 11]
```diff
@@ -124,6 +124,7 @@ FORCE_4BY3 = "Wymus 4:3"
 PRELOAD_TEXTURES = "Wstepnie Zaladuj Tekstury"
 VSYNC = "Synchronizacja Pionowa"
 UNCAPPED_FRAMERATE = "Nieograniczona Liczba Klatek"
+REDUCE_FRAMERATE_FOCUS_LOSS = "Zmniejsz liczbę klatek na sekundę, gdy okno jest nieostre"
```
Comment (Flower35) [inline lang/Polish.ini]: Please don't just put a text through a translator and call it a day.

I know this is mainly a code-change Pull Request, but it ends up having poor quality and bringing bad gaming experience with unclear messages.

Here *"Window Unfocused"* got translated as *„nieostre”* with the meaning closer to *"unsharp"*, *"blurry"*.

Correct translation would be either:
- *„gdy okno traci fokus”* (*"when the window loses focus"*),
- or *„gdy okno nie ma fokusa”* (*"when the window is unfocused"*),
- or *„gdy okno jest nieaktywne”* ("when the window is inactive").

Comment (retcinder) [reply lang/Polish.ini]: > Please don't just put a text through a translator and call it a day.
> 
> I know this is mainly a code-change Pull Request, but it ends up having poor quality and bringing bad gaming experience with unclear messages.
> 

Yeah, sorry about that. I'll look for native/fluent speakers of the non-English languages to ensure the text is replaced with ones that actually make sense (German has been done like that so far, currently getting Spanish done).

> Correct translation would be either:
> 
>     * _„gdy okno traci fokus”_ (_"when the window loses focus"_),
> 
>     * or _„gdy okno nie ma fokusa”_ (_"when the window is unfocused"_),
> 
>     * or _„gdy okno jest nieaktywne”_ ("when the window is inactive").

First one seems seems good. Thanks!

**Comment by tulpenkiste** [comment]
PR is almost ready for review, waiting for a dutch speaker to get back to me and then it'll be complete.

File: lang/Spanish.ini (hunk 6 above)
Comment (retcinder) [reply lang/Spanish.ini]: Retranslated, new text fits in panel

File: lang/Spanish.ini (hunk 5 above)
Comment (retcinder) [reply lang/Spanish.ini]: Retranslated, new text fits in panel

**Comment by tulpenkiste** [comment]
All non-english translations were replaced with the help of some speakers of the respective languages.
All fit in their respective menus.

**Comment by AgentXLP** [comment]
yes, sorry I didn't comment it I got sidetracked closing other issues and pull requests.
I am currently porting PRs to the repo that sm64coopdx 1.0 is being worked on, I am likely going to port your code and then touch it up and co-author you in the commit so you get credit properly.

**Pull request closed without merging**